src/function.cpp

```diff
@@ -59,32 +55,82 @@ DeclaredFunction DeclaredFunction::declare(std::vector<mlir::Type> &&domain,
   transform(domain.cbegin(), domain.cend(), back_inserter(smtDomain),
             typeToSort);
   FnDecl decl(smtDomain, typeToSort(range), string(name) + "_tvfn");
-  return DeclaredFunction(move(domain), move(range), move(decl));
+
+  vector<FnDecl> dims;
+  if (auto sty = range.dyn_cast<mlir::ShapedType>()) {
+    const auto rank = sty.getRank();
+    dims.reserve(rank);
+    const auto dimPrefix = string(name) + "_tvfn_dim_";
+    for (size_t i = 0; i < rank; i++) {
+      auto dim = FnDecl(smtDomain, Index::sort(), dimPrefix + to_string(i));
+      dims.push_back(move(dim));
+    }
+  }
+  return DeclaredFunction(move(domain), move(range), move(decl), move(dims));
 }
 
 ValueTy DeclaredFunction::apply(const std::vector<ValueTy> &operands) const {
   vector<Expr> operandExprs;
   operandExprs.reserve(operands.size());
 
+  auto getZeroGuardedExpr = [](const ValueTy &val) {
+    if (holds_alternative<Tensor>(val)) {
+      const auto tensorVal = get<Tensor>(val);
+      Expr numElements = Index::one();
+      for (const auto &dim : tensorVal.getDims()) {
+        numElements = numElements * dim;
+      }
+
+      const Expr arr = tensorVal;
+      const auto i = static_cast<Expr>(Index::var("idx", VarType::BOUND));
+
+      Expr zero = Index::zero();
+      const auto elemType = tensorVal.getElemType();
+      if (elemType.isa<mlir::FloatType>()) {
+        zero = Float::constant(llvm::APFloat(0.0), elemType);
+      } else if (elemType.isa<mlir::IntegerType>()) {
+        zero = Integer(0, elemType.getIntOrFloatBitWidth());
```

Comment:
I remember there was something like `getZero(ty)`... could you check?

Reply:
We do have `llvm::APFloat::getZero(llvm::Semantics)` but specifying the semantics require several more lines of code. Would it still be better to use getZero?

Reply:
No, there is one is at `value.h`; maybe we can use that.

Reply:
Fixed